sqlccl: cleanup failed or canceled RESTORE data

madelynnblue opened the pull request:
Previously, if a RESTORE failed or was manually canceled, any data
it had committed would have been orphaned, and forever taken up
space. Since there was no table descriptor, nothing would have been
able to see or delete it.

Add optional callbacks in jobs code to allow for specific jobs
to implement an onFail method. Add the job's client.Txn to the
update method and this onFail callback to allow for a job to be
transactionally failed and cleaned up.

Release note: Cleanup partially restored data when a RESTORE fails
or is canceled.

Fixes #19398
Fixes #17123

@vivekmenezes could you review the table drop stuff in restore.go?

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19578)
<!-- Reviewable:end -->

benesch commented on pkg/ccl/sqlccl/backup_test.go:
```diff
@@ -864,9 +864,53 @@ func TestBackupRestoreControlJob(t *testing.T) {
+	}
+	// Verify the failed RESTORE added some DROP tables.
+	sqlDB.CheckQueryResults(
+		`SELECT name FROM crdb_internal.tables WHERE database_name = 'restore' AND state = 'DROP'`,
```
I worry this might be flaky. What's to stop the schema changer from immediately cleaning up the failed backup, before this query runs? Maybe there's something we can do with `SchemaChangerTestingKnobs.AsyncExecNotification`.

I suppose if stress/stressrace seem happy I'm not particularly fussed.

benesch commented on pkg/sql/jobs/jobs.go:
```diff
@@ -276,6 +277,11 @@ func (j *Job) Canceled(ctx context.Context) error {
+			if err := onfail.onFail(ctx, txn, j); err != nil {
+				return false, err
+			}
+		}
```
👍

benesch commented on pkg/ccl/sqlccl/restore.go:
```diff
@@ -1153,6 +1153,52 @@ func doRestorePlan(
+// in DROP state, which causes the schema change stuff to delete the keys
+// in the background.
+func restoreFailHook(ctx context.Context, txn *client.Txn, settings *cluster.Settings, details *jobs.RestoreDetails) error {
+	// Needed to trigger the schema change manager.
```
More precisely it's needed to trigger a re-gossiping of the system config. Any transaction that touches the system config span is supposed to set this, which I'm realizing we don't do elsewhere in restore. I've filed #19617 to track, because `SetSystemConfigTrigger` confuses me every time I see it.

benesch reviewed (commented):
Your approach looks entirely perfect—just see my one little note about potential flakiness in the test.

benesch commented on pkg/sql/jobs/jobs.go:
```diff
@@ -308,11 +319,26 @@ func (j *Job) Failed(ctx context.Context, err error) {
+
+func (r *Payload_Restore) onFail(ctx context.Context, txn *client.Txn, job *Job) error {
+	// This is set in the CCL code if RESTOREs are enabled, so no need to check
+	// for nil.
```
Heh, looks like this isn't true when running the tests for the jobs package. 🤦‍♂️

mjibson commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 909 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578#-KxPMcHrLk9MFgxPJAnE-r1-909:-Kxg70sMsdzWi3OS_eri:bn6111w) ([raw file](https://github.com/cockroachdb/cockroach/blob/89b5673d8d5f7a8019184e990e1a164c095234e8/pkg/ccl/sqlccl/backup_test.go#L909)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I worry this might be flaky. What's to stop the schema changer from immediately cleaning up the failed backup, before this query runs? Maybe there's something we can do with `SchemaChangerTestingKnobs.AsyncExecNotification`.

I suppose if stress/stressrace seem happy I'm not particularly fussed.
</blockquote></details>

A stress didn't show this, but I've added it anyway just in case. Good eye.

---

*[pkg/ccl/sqlccl/restore.go, line 1177 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578#-KxPMcHrLk9MFgxPJAnF-r1-1177:-Kxg7DhMuEBQFFIOHj8T:b-2rsho) ([raw file](https://github.com/cockroachdb/cockroach/blob/89b5673d8d5f7a8019184e990e1a164c095234e8/pkg/ccl/sqlccl/restore.go#L1177)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

More precisely it's needed to trigger a re-gossiping of the system config. Any transaction that touches the system config span is supposed to set this, which I'm realizing we don't do elsewhere in restore. I've filed #19617 to track, because `SetSystemConfigTrigger` confuses me every time I see it.
</blockquote></details>

Ah, TIL. I had no idea about this until Vivek told me it was causing the lack of processing.

---

*[pkg/sql/jobs/jobs.go, line 333 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578#-KxPMcHrLk9MFgxPJAnG-r1-333:-Kxg83TIH_qWU9YOonIS:b-z0othp) ([raw file](https://github.com/cockroachdb/cockroach/blob/89b5673d8d5f7a8019184e990e1a164c095234e8/pkg/sql/jobs/jobs.go#L333)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Heh, looks like this isn't true when running the tests for the jobs package. 🤦‍♂️ 
</blockquote></details>

Oh indeed. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578)*
<!-- Sent from Reviewable.io -->

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup_test.go, line 797 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578#-Kxhy7ho_dWDJu0y7Zfa:-Kxhy7ho_dWDJu0y7Zfb:b-a3ito) ([raw file](https://github.com/cockroachdb/cockroach/blob/71c1a1e51d0f9b99be35615f36771cbb5588cd97/pkg/ccl/sqlccl/backup_test.go#L797)):*
> ```Go
> 	// enough that it will generate more than one export or import response.
> 	var allowResponse chan struct{}
> 	params := base.TestClusterArgs{}
> ```

I think you want the schema change knob here, too?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578#-:-KxhyBYDnfnV8vPs1prz:bnfp4nl)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sqlccl: cleanup failed or canceled RESTORE data

Previously, if a RESTORE failed or was manually canceled, any data
it had committed would have been orphaned, and forever taken up
space. Since there was no table descriptor, nothing would have been
able to see or delete it.

Add optional callbacks in jobs code to allow for specific jobs
to implement an onFail method. Add the job's client.Txn to the
update method and this onFail callback to allow for a job to be
transactionally failed and cleaned up.

Release note: Cleanup partially restored data when a RESTORE fails
or is canceled.

Fixes #19398
Fixes #17123

mjibson commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup_test.go, line 797 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578#-Kxhy7ho_dWDJu0y7Zfa:-KxirrKy_FibTgAiRxVb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71c1a1e51d0f9b99be35615f36771cbb5588cd97/pkg/ccl/sqlccl/backup_test.go#L797)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think you want the schema change knob here, too?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19578)*
<!-- Sent from Reviewable.io -->
